{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Lab_Exercise_12.ipynb","provenance":[],"collapsed_sections":[],"mount_file_id":"14WFyv-ASJ5voGnCnh_hlkDgLWa_Iy3mm","authorship_tag":"ABX9TyPHvym/rl91CNvBKWL7PDiD"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"ubkHAESo-uIY"},"outputs":[],"source":[""]},{"cell_type":"code","execution_count":null,"metadata":{"id":"D_LM3B2s1P0x"},"outputs":[],"source":["!pip install biopython"]},{"cell_type":"code","source":["!pip install PyVCF"],"metadata":{"id":"bcxANobJ2SlU","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1646700313049,"user_tz":-60,"elapsed":12724,"user":{"displayName":"L bio-TIC","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gj0_9O8e326BUqjNr-3ykzASjdb2MCzdm9_0_iiaA=s64","userId":"05412660529920510610"}},"outputId":"1aab35b5-331b-4c3b-c6f0-613197e6006a"},"execution_count":1,"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting PyVCF\n","  Downloading PyVCF-0.6.8.tar.gz (34 kB)\n","Requirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from PyVCF) (57.4.0)\n","Building wheels for collected packages: PyVCF\n","  Building wheel for PyVCF (setup.py) ... \u001b[?25l\u001b[?25hdone\n","  Created wheel for PyVCF: filename=PyVCF-0.6.8-cp37-cp37m-linux_x86_64.whl size=124145 sha256=23e4038d182aac47cff917930eff5232d5696c604342b1feb20060dd03f138e7\n","  Stored in directory: /root/.cache/pip/wheels/8a/60/8d/fe98f8401a0bb73b44afbbc454d1d7fe11a0a2081f5b899597\n","Successfully built PyVCF\n","Installing collected packages: PyVCF\n","Successfully installed PyVCF-0.6.8\n"]}]},{"cell_type":"code","execution_count":2,"metadata":{"id":"SeBA7GSD-je2","executionInfo":{"status":"ok","timestamp":1646700316671,"user_tz":-60,"elapsed":1136,"user":{"displayName":"L bio-TIC","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gj0_9O8e326BUqjNr-3ykzASjdb2MCzdm9_0_iiaA=s64","userId":"05412660529920510610"}}},"outputs":[],"source":["import os\n","import numpy as np\n","import pandas as pd\n","import matplotlib.pyplot as plt\n","import vcf\n","from IPython.display import display\n","from ipywidgets import FloatProgress\n","%matplotlib inline\n","plt.style.use('ggplot')"]},{"cell_type":"code","execution_count":3,"metadata":{"scrolled":true,"outputId":"7fd7b0af-1bc4-4aa0-c626-4bcf067cc2ec","colab":{"base_uri":"https://localhost:8080/","height":424},"id":"-aAfpIXY_PJj","executionInfo":{"status":"ok","timestamp":1646700333298,"user_tz":-60,"elapsed":651,"user":{"displayName":"L bio-TIC","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gj0_9O8e326BUqjNr-3ykzASjdb2MCzdm9_0_iiaA=s64","userId":"05412660529920510610"}}},"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <div id=\"df-eb45a37e-1bbf-44c1-ad50-486b9a4da6d5\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>HGVS Variant</th>\n","      <th>Pathogenicity</th>\n","      <th>Pathogenic</th>\n","      <th>Start</th>\n","      <th>End</th>\n","      <th>Reference Seq</th>\n","      <th>Variant Seq</th>\n","      <th>Reference Seq Complement</th>\n","      <th>Variant Seq Complement</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>c.8T&gt;G</td>\n","      <td>5</td>\n","      <td>True</td>\n","      <td>41276105</td>\n","      <td>41276105</td>\n","      <td>A</td>\n","      <td>C</td>\n","      <td>T</td>\n","      <td>G</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>c.32_33insC</td>\n","      <td>5</td>\n","      <td>True</td>\n","      <td>41276080</td>\n","      <td>41276081</td>\n","      <td>TA</td>\n","      <td>TGA</td>\n","      <td>TA</td>\n","      <td>TCA</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>c.34C&gt;T</td>\n","      <td>5</td>\n","      <td>True</td>\n","      <td>41276079</td>\n","      <td>41276079</td>\n","      <td>G</td>\n","      <td>A</td>\n","      <td>C</td>\n","      <td>T</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>c.38_39delATinsGGG</td>\n","      <td>5</td>\n","      <td>True</td>\n","      <td>41276074</td>\n","      <td>41276075</td>\n","      <td>AT</td>\n","      <td>CCC</td>\n","      <td>AT</td>\n","      <td>GGG</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>c.53T&gt;C</td>\n","      <td>4</td>\n","      <td>True</td>\n","      <td>41276060</td>\n","      <td>41276060</td>\n","      <td>A</td>\n","      <td>G</td>\n","      <td>T</td>\n","      <td>C</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>1163</th>\n","      <td>c.5553dupC</td>\n","      <td>5</td>\n","      <td>True</td>\n","      <td>41197733</td>\n","      <td>41197733</td>\n","      <td>G</td>\n","      <td>GG</td>\n","      <td>C</td>\n","      <td>CC</td>\n","    </tr>\n","    <tr>\n","      <th>1164</th>\n","      <td>c.5558dupA</td>\n","      <td>5</td>\n","      <td>True</td>\n","      <td>41197728</td>\n","      <td>41197728</td>\n","      <td>T</td>\n","      <td>TT</td>\n","      <td>A</td>\n","      <td>AA</td>\n","    </tr>\n","    <tr>\n","      <th>1165</th>\n","      <td>c.5559C&gt;G</td>\n","      <td>5</td>\n","      <td>True</td>\n","      <td>41197727</td>\n","      <td>41197727</td>\n","      <td>G</td>\n","      <td>C</td>\n","      <td>C</td>\n","      <td>G</td>\n","    </tr>\n","    <tr>\n","      <th>1166</th>\n","      <td>c.5572A&gt;C</td>\n","      <td>1</td>\n","      <td>False</td>\n","      <td>41197714</td>\n","      <td>41197714</td>\n","      <td>T</td>\n","      <td>G</td>\n","      <td>A</td>\n","      <td>C</td>\n","    </tr>\n","    <tr>\n","      <th>1167</th>\n","      <td>c.5576C&gt;G</td>\n","      <td>1</td>\n","      <td>False</td>\n","      <td>41197710</td>\n","      <td>41197710</td>\n","      <td>G</td>\n","      <td>C</td>\n","      <td>C</td>\n","      <td>G</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>1168 rows Ã— 9 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-eb45a37e-1bbf-44c1-ad50-486b9a4da6d5')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-eb45a37e-1bbf-44c1-ad50-486b9a4da6d5 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-eb45a37e-1bbf-44c1-ad50-486b9a4da6d5');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "],"text/plain":["            HGVS Variant  Pathogenicity  Pathogenic     Start       End  \\\n","0                 c.8T>G              5        True  41276105  41276105   \n","1            c.32_33insC              5        True  41276080  41276081   \n","2                c.34C>T              5        True  41276079  41276079   \n","3     c.38_39delATinsGGG              5        True  41276074  41276075   \n","4                c.53T>C              4        True  41276060  41276060   \n","...                  ...            ...         ...       ...       ...   \n","1163          c.5553dupC              5        True  41197733  41197733   \n","1164          c.5558dupA              5        True  41197728  41197728   \n","1165           c.5559C>G              5        True  41197727  41197727   \n","1166           c.5572A>C              1       False  41197714  41197714   \n","1167           c.5576C>G              1       False  41197710  41197710   \n","\n","     Reference Seq Variant Seq Reference Seq Complement Variant Seq Complement  \n","0                A           C                        T                      G  \n","1               TA         TGA                       TA                    TCA  \n","2                G           A                        C                      T  \n","3               AT         CCC                       AT                    GGG  \n","4                A           G                        T                      C  \n","...            ...         ...                      ...                    ...  \n","1163             G          GG                        C                     CC  \n","1164             T          TT                        A                     AA  \n","1165             G           C                        C                      G  \n","1166             T           G                        A                      C  \n","1167             G           C                        C                      G  \n","\n","[1168 rows x 9 columns]"]},"metadata":{}}],"source":["brca1_variants = pd.read_csv(os.path.join('/content/drive/MyDrive/Quantitative_Biological_Research_with_Python/VCF_Multivar_Analysis/brca1_variants.csv'))\n","display(brca1_variants)"]},{"cell_type":"code","execution_count":4,"metadata":{"id":"Ucb5YCmY-jfC","outputId":"c8e915ae-585c-4952-bfc7-05e91d468082","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1646700341240,"user_tz":-60,"elapsed":279,"user":{"displayName":"L bio-TIC","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gj0_9O8e326BUqjNr-3ykzASjdb2MCzdm9_0_iiaA=s64","userId":"05412660529920510610"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["1085\n","[(41244887, 'GC', 'GTC'), (41226449, 'GAGA', nan), (41243478, 'CTTG', nan), (41244413, 'TTAATATTGC', nan), (41243854, 'A', 'AA')]\n"]}],"source":["# We hold a cache of all pathogenic variants in a set (for fast lookup). Each variant is represented as a tuple of:\n","# (location (0-based index; int), ref seq (str), alt seq (str)).\n","\n","pathogenic_variants = set(brca1_variants[brca1_variants['Pathogenic']].apply(lambda record: \\\n","        (record['Start'], record['Reference Seq'], record['Variant Seq']), axis = 1))\n","\n","print(len(pathogenic_variants))\n","print(list(pathogenic_variants)[:5])"]},{"cell_type":"code","execution_count":5,"metadata":{"id":"rLjxC7HK-jfE","outputId":"7b3a7c77-083f-47e2-8db2-ab25f7162305","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1646700434694,"user_tz":-60,"elapsed":37506,"user":{"displayName":"L bio-TIC","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gj0_9O8e326BUqjNr-3ykzASjdb2MCzdm9_0_iiaA=s64","userId":"05412660529920510610"}}},"outputs":[{"output_type":"stream","name":"stdout","text":["Sample HG00403 has the variant T --> A at position 41201105 (rs80358092)\n","Sample HG00537 has the variant T --> A at position 41201105 (rs80358092)\n","Sample HG00626 has the variant T --> A at position 41201105 (rs80358092)\n","Sample NA18528 has the variant T --> A at position 41201105 (rs80358092)\n","Sample NA18570 has the variant T --> A at position 41201105 (rs80358092)\n","Sample NA18613 has the variant T --> A at position 41201105 (rs80358092)\n","Sample NA18641 has the variant T --> A at position 41201105 (rs80358092)\n","Sample NA19780 has the variant A --> C at position 41243840 (rs28897687)\n","Sample NA19380 has the variant C --> G at position 41243948 (rs56214134)\n","Sample NA19382 has the variant C --> G at position 41243948 (rs56214134)\n"]}],"source":["VCF_FILE_NAME = '/content/drive/MyDrive/Quantitative_Biological_Research_with_Python/VCF_Multivar_Analysis/ALL.chr17.integrated_phase1_v3.20101123.snps_indels_svs.genotypes.41100000-41300000.vcf.gz'\n","\n","for variant in vcf.Reader(filename = os.path.join(VCF_FILE_NAME), compressed = True):\n","    for allele_index, alt in enumerate(variant.ALT):\n","        \n","        # variant_key will have the same representation as the elements in the lookup set of pathogenic variants.\n","        # Note that we convert the position from 1-based to 0-based index, and that an empty sequence is represented as np.nan.\n","        variant_key = (variant.POS - 1, str(variant.REF), np.nan if len(alt) == 0 else str(alt))\n","        \n","        if variant_key in pathogenic_variants:\n","            for call in variant.samples:\n","                # call.gt_alleles gives the genotypes of the current sample as numbers. 0 represents the ref allele, 1 represents\n","                # the first alternative allele, and so on. On the other hand, allele_index starts counting the first alternative\n","                # allele as 0. Therefore, we check whether (allele_index + 1) is in the list of allele numbers of the genotype\n","                # of the current sample. \n","                if (allele_index + 1) in map(int, call.gt_alleles):\n","                    print('Sample %s has the variant %s --> %s at position %d (%s)' % \\\n","                            (call.sample, variant.REF, alt, variant.POS, variant.ID))"]}]}